abci/checktx/mempool_parity_check_tx.go

```diff
@@ -72,6 +74,28 @@ func (m MempoolParityCheckTx) CheckTx() CheckTx {
 		}
 
 		// run the checkTxHandler
-		return m.checkTxHandler(req)
+		res, checkTxError := m.checkTxHandler(req)
+
+		// if re-check fails for a transaction, we'll need to explicitly purge the tx from
+		// the app-side mempool
+		if isInvalidCheckTxExecution(res, checkTxError) && isReCheck {
+			// check if the tx exists first
+			if m.mempl.Contains(tx) {
+				// remove the tx
+				if err := m.mempl.Remove(tx); err != nil {
+					m.logger.Debug(
+						"failed to remove tx from app-side mempool when purging for re-check failure",
+						"removal-err", err,
+						"check-tx-err", checkTxError,
+					)
+				}
+			}
+		}
+
+		return res, checkTxError
 	}
 }
+
+func isInvalidCheckTxExecution(resp *cmtabci.ResponseCheckTx, checkTxErr error) bool {
+	return resp == nil || resp.Code != 0 || checkTxErr != nil
```

Comment:
Seems fine. Not sure if we should prefer the [IsErr](https://github.com/cometbft/cometbft/blob/6da2cf396c145081ab534cd9b72810a531812a14/api/cometbft/abci/v1beta3/types.go#L19) function, but overall assuming nonzero=failure seems correct.